server/src/main/java/org/elasticsearch/search/sort/RawBucketedSort.java

```diff
@@ -0,0 +1,340 @@
+        long rootIndex = bucket * bucketSize + 1;
+        if (inHeapMode(bucket)) {
+            if (betterThan(0, rootIndex)) {
+                // Insert the temporal value into the heap
+                swap(0, rootIndex);
+                downHeap(rootIndex, 0);
+            }
+            return;
+        }
+        // Gathering mode
+        long requiredSize = rootIndex + bucketSize;
+        if (values().size() < requiredSize) {
+            grow(requiredSize);
+        }
+        int next = getNextGatherOffset(rootIndex);
+        assert 0 <= next && next < bucketSize
+            : "Expected next to be in the range of valid buckets [0 <= " + next + " < " + bucketSize + "]";
+        long index = next + rootIndex;
+        swap(0, index);
+        if (next == 0) {
+            heapMode.set(bucket);
+            heapify(rootIndex);
+        } else {
+            setNextGatherOffset(rootIndex, next - 1);
+        }
+    }
+
+    /**
+     * Superclass for implementations of {@linkplain RawBucketedSort} for {@code long} keys.
+     */
+    public static class ForLongs extends RawBucketedSort<Long> {
+        private LongArray values;
+
+        @SuppressWarnings("this-escape")
+        public ForLongs(BigArrays bigArrays, SortOrder sortOrder, int bucketSize) {
+            super(bigArrays, sortOrder, bucketSize);
+            boolean success = false;
+            try {
+                values = bigArrays.newLongArray(1, false);
+                success = true;
+            } finally {
+                if (success == false) {
+                    close();
+                }
+            }
+        }
+
+        @Override
+        protected final BigArray values() {
+            return values;
+        }
+
+        @Override
+        protected final void growValues(long minSize) {
+            values = bigArrays.grow(values, minSize);
+        }
+
+        @Override
+        protected final int getNextGatherOffset(long rootIndex) {
+            return (int) values.get(rootIndex);
+        }
+
+        @Override
+        protected final void setNextGatherOffset(long rootIndex, int offset) {
+            values.set(rootIndex, offset);
+        }
+
+        @Override
+        protected final boolean betterThan(long lhs, long rhs) {
+            return getOrder().reverseMul() * Long.compare(values.get(lhs), values.get(rhs)) < 0;
+        }
+
+        @Override
+        protected final void swap(long lhs, long rhs) {
+            long tmp = values.get(lhs);
+            values.set(lhs, values.get(rhs));
+            values.set(rhs, tmp);
+        }
+
+        // Type-specific methods
+
+        /**
+         * Collects a value into a bucket.
+         */
+        public void collect(long bucket, long value) {
+            values.set(0, value);
+            super.collect(bucket);
+        }
+
+        @Override
+        public List<Long> getValues(long bucket) {
+            var bounds = getBucketValuesBounds(bucket);
```

Comment:
I think this method will probably want to change to `Block getValues(int[] buckets)` - that way we can do everything `Block`y.